futures-util/src/try_stream/try_zip.rs

```diff
@@ -0,0 +1,122 @@
+    queued1: Option<Result<St1::Ok, St1::Error>>,
+    queued2: Option<Result<St2::Ok, St2::Error>>,
+}
+
+impl<St1: TryStream + Unpin, St2: TryStream + Unpin> Unpin for TryZip<St1, St2> {}
+
+impl<St1: TryStream, St2: TryStream> TryZip<St1, St2> {
+    unsafe_pinned!(stream1: Fuse<IntoStream<St1>>);
+    unsafe_pinned!(stream2: Fuse<IntoStream<St2>>);
+    unsafe_unpinned!(queued1: Option<Result<St1::Ok, St1::Error>>);
+    unsafe_unpinned!(queued2: Option<Result<St2::Ok, St2::Error>>);
+
+    pub(super) fn new(stream1: St1, stream2: St2) -> TryZip<St1, St2> {
+        TryZip {
+            stream1: stream1.into_stream().fuse(),
+            stream2: stream2.into_stream().fuse(),
+            queued1: None,
+            queued2: None,
+        }
+    }
+
+    /// Acquires a reference to the underlying streams that this combinator is
+    /// pulling from.
+    pub fn get_ref(&self) -> (&St1, &St2) {
+        (self.stream1.get_ref().get_ref(), self.stream2.get_ref().get_ref())
+    }
+
+    /// Acquires a mutable reference to the underlying streams that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_mut(&mut self) -> (&mut St1, &mut St2) {
+        (self.stream1.get_mut().get_mut(), self.stream2.get_mut().get_mut())
+    }
+
+    /// Acquires a pinned mutable reference to the underlying streams that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_pin_mut<'a>(self: Pin<&'a mut Self>) -> (Pin<&'a mut St1>, Pin<&'a mut St2>)
+        where St1: Unpin, St2: Unpin,
+    {
+        let Self { stream1, stream2, .. } = Pin::get_mut(self);
+        (Pin::new(stream1.get_mut().get_mut()), Pin::new(stream2.get_mut().get_mut()))
+    }
+
+    /// Consumes this combinator, returning the underlying streams.
+    ///
+    /// Note that this may discard intermediate state of this combinator, so
+    /// care should be taken to avoid losing resources when this is called.
+    pub fn into_inner(self) -> (St1, St2) {
+        (self.stream1.into_inner().into_inner(), self.stream2.into_inner().into_inner())
+    }
+}
+
+impl<St1, St2> FusedStream for TryZip<St1, St2>
+    where St1: TryStream, St2: TryStream,
+{
+    fn is_terminated(&self) -> bool {
+        self.stream1.is_terminated() && self.stream2.is_terminated()
+    }
+}
+
+impl<St1, St2> TryStream for TryZip<St1, St2>
+where
+    St1: TryStream,
+    St2: TryStream<Error = St1::Error>
+{
+    type Ok = (St1::Ok, St2::Ok);
+    type Error = St1::Error;
+
+    fn try_poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Ok, Self::Error>>> {
+        if self.queued1.is_none() {
+            match self.as_mut().stream1().try_poll_next(cx) {
```

Comment:
I think using `self.as_mut().stream1().try_poll_next(cx)?` here will avoid the problem you mentioned.

Reply:
Wouldn't that cause the two halves of the zipper to become unaligned? That is,

```
A: Ok(1), Err(2), Ok(3)
B: Ok(4), Ok(5), Ok(6)
```

Would result in 

```
Ok((1, 4)), Err(2), Ok((3,5))
```

Is that desired? (Perhaps this also suggests why this combinator hasn't bee created before now).

Reply:
Hmm, but that's the same behavior as `Stream::zip` of futures 0.1.

https://github.com/rust-lang-nursery/futures-rs/blob/5654784d5b677dadc4b86affc575d12aa737e4a7/src/stream/zip.rs#L37

Reply:
> the same behavior as `Stream::zip` of futures 0.1

It sure is:

```rust
use futures::{future, stream, Future, Stream}; // 0.1.26

fn main() {
    let a: Vec<Result<i32, i32>> = vec![Ok(1), Err(2), Ok(3)];
    let b: Vec<Result<i32, i32>> = vec![Ok(4), Ok(5), Ok(6)];

    stream::iter_result(a)
        .zip(stream::iter_result(b))
        .then(|x| future::ok::<_, ()>(eprintln!("{:?}", x)))
        .for_each(|_| Ok(()))
        .wait();
}
```

([Playground](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=bbedf213c602be358159fce4f6ca5493))

Output:

```
Ok((1, 4))
Err(2)
Ok((3, 5))
```

But I think the question remains: is this the expected and desired behavior?

Reply:
> is this the expected and desired behavior?

If we are open to changing behaviors, the most comprehensive solution is to have

```rust
enum ZipError<A, B> {
    A(A),
    B(B),
    Both(A, B)
}
```

Reply:
Hmm... I have no clear opinion on which is better, for now.

FWIW, as another thing we may need to consider, the behavior of current this PR's TryZip can be written using Zip, but the behavior of 0.1's Zip cannot be written using either 0.3's Zip or current this PR's TryZip.

cc @cramertj @Nemo157 Do you have any thoughts?

Reply:
I would've personally expected that `try_zip` would stop returning at all after the first error, but I don't have a strong opinion here.

Reply:
Bump-- @shepmaster what are your thoughts here?

Reply:
Sadly, I don't even remember what my original use case was for this, and I don't recall what my thoughts were. 😊 

I've posted the open questions back on the issue. The fact that there aren't people clamoring for this feature is perhaps a sign that it's not useful, yet.

Reply:
Okay! Let me know if you're interested in revisiting at some point.